db/collection.go

```diff
@@ -41,16 +43,39 @@ type Collection struct {
 	indexes      map[string]Index
 }
 
-func newCollection(name string, schema string, d *DB) *Collection {
-	schemaLoader := gojsonschema.NewStringLoader(schema)
+func newCollection(name string, schema *jsonschema.Schema, d *DB) (*Collection, error) {
+	// by default, use top level properties to validate ID string property exists
+	properties := schema.Properties
+	if schema.Ref != "" {
+		// the schema specifies a ref to a nested type, use it instead
+		parts := strings.Split(schema.Ref, "/")
+		if len(parts) < 1 {
+			return nil, ErrInvalidCollectionSchema
+		}
+		typeName := parts[len(parts)-1]
+		refDefinition := schema.Definitions[typeName]
+		if refDefinition == nil {
+			return nil, ErrInvalidCollectionSchema
+		}
+		properties = refDefinition.Properties
+	}
+	if !hasIDProperty(properties) {
+		return nil, ErrInvalidCollectionSchema
+	}
```

Comment:
Above is the schema validation. As far as I can tell, there are two ways an `ID` might be defined, and we have to handle them both. First is the "expanded" schema case where a type is defined in the top level of the schema. Second is when the top level schema specifies a `Ref` to type nested in the schema `Definitions`.

Reply:
I still like the idea of validating the schema against another schema, but this was a more clear approach for now.